### PR TITLE
Add service to Achievement github repository

### PR DESCRIPTION
**Pull request by VladSpitsin**
* add solution "LT.DigitalOffice.AchievementService";
* add directory "Solution items" and put the following files there: ".dockerignore", ".editorconfig", ".gitignore", "Common.Configurations.targets", "Dockerfile", "main.yml", "README.md"
* add directory "src" and put the following items:
application "AchievementService" (add appsettings.json, set Program.cs, Startup.cs), add following DLLs: AchievementService.Broker (add "Consumers" directory), AchievementService.Business (add "Commands" directory), AchievementService.Data (add "Interfaces" directory), AchievementService.Data.Provider (add "IDataProvider" interface), AchievementService.Data.Provider.MsSql.Ef (add "Migrations" directory and "AchievementServiceDbContext" class), AchievementService.Mappers, AchievementService.Models.Db, AchievementService.Models.Dto, AchievementService.Validation

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+RUN dotnet restore -s https://api.nuget.org/v3/index.json
+
+COPY . ./
+RUN dotnet publish -c Release -o out
+
+FROM mcr.microsoft.com/dotnet/aspnet:5.0-buster-slim AS base
+WORKDIR /app
+COPY --from=build /app/out .
+EXPOSE 80
+ENTRYPOINT ["dotnet", "LT.DigitalOffice.CompanyService.dll"]
```
Comment (ekaterinaryzhenkova) [inline Dockerfile]: suppose AchieveService.dll

File: src/AchievementService/Properties/launchSettings.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  "iisSettings": {
+    "windowsAuthentication": false,
+    "anonymousAuthentication": true,
+    "iisExpress": {
+      "applicationUrl": "http://localhost:9104",
+      "sslPort": 0
+    }
+  },
+  "profiles": {
+    "IIS Express": {
```
Comment (ekaterinaryzhenkova) [inline src/AchievementService/Properties/launchSettings.json]: do you need this block?

File: src/AchievementService/AchievementService.csproj [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
```
Comment (actimellka) [inline src/AchievementService/AchievementService.csproj:2]: maybe you need  <Import Project="$(ProjectDir)..\..\Common.Configurations.targets" /> here and in libraries, consult Mariya

Comment (actimellka) [inline src/AchievementService/AchievementService.csproj:2]: maybe you need "<Import Project="$(ProjectDir)..\..\Common.Configurations.targets" />" here and in libraries, consult Mariya

Comment (ekaterinaryzhenkova) [inline src/AchievementService/AchievementService.csproj:2]: maybe you need a connection with CommonConfigurationTargets here and in libraries, consult Mariya

File: src/AchievementService/Properties/launchSettings.json (hunk 2 above)
Comment (VladSpitsin) [reply src/AchievementService/Properties/launchSettings.json]: fixed

File: src/AchievementService.Data.Provider/IDataProvider.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace AchievementService.Data.Provider
+{
+    public interface IDataProvider
```
Comment (Cheephot) [inline src/AchievementService.Data.Provider/IDataProvider.cs]: [AutoInject(InjectType.Scoped)]

Comment (asyagraf) [inline src/AchievementService.Data.Provider/IDataProvider.cs]: IDataProvider inherits IBaseDataProvider

File: src/AchievementService/Program.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.AchievementService
+{
+    public class Program
```
Comment (asyagraf) [inline src/AchievementService/Program.cs]: fix tabs

File: src/AchievementService/Startup.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+
+namespace LT.DigitalOffice.AchievementService
+{
```
Comment (asyagraf) [inline src/AchievementService/Startup.cs:35]: fix tabs

File: src/AchievementService/appsettings.Development.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (asyagraf) [inline src/AchievementService/appsettings.Development.json]: don't you need anything else here? mb you should check another services or consult Maria

File: src/AchievementService/appsettings.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+  "Logging": {
```
Comment (asyagraf) [inline src/AchievementService/appsettings.json]: same

File: src/AchievementService.Data.Provider.MsSql.Ef/AchievementServiceDbContext.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+
+namespace AchievementService.Data.Provider.MsSql.Ef
+{
+    public class AchievementServiceDbContext : DbContext, IDataProvider
```
Comment (asyagraf) [inline src/AchievementService.Data.Provider.MsSql.Ef/AchievementServiceDbContext.cs]: Add methods of IDataProvider, you can use AdminService as example

File: src/AchievementService/AchievementService.csproj [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    <PackageReference Include="Serilog.AspNetCore" Version="4.1.0" />
+    <PackageReference Include="Serilog.Sinks.Seq" Version="5.1.0" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\AchievementService.Broker\AchievementService.Broker.csproj" />
+    <ProjectReference Include="..\AchievementService.Business\AchievementService.Business.csproj" />
+    <ProjectReference Include="..\AchievementService.Models.Dto\AchievementService.Models.Dto.csproj" />
+    <ProjectReference Include="..\AchievementService.Validation\AchievementService.Validation.csproj" />
+    <ProjectReference Include="..\AchievementService.Broker\AchievementService.Broker.csproj" />
```
Comment (MariyaSin) [inline src/AchievementService/AchievementService.csproj]: the ref exact like a 14 line

Comment (MariyaSin) [reply src/AchievementService/AchievementService.csproj]: and next lines have same problem

File: src/AchievementService/Program.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+      string seqApiKey = Environment.GetEnvironmentVariable("seqApiKey");
+      if (string.IsNullOrEmpty(seqApiKey))
+      {
+        seqApiKey = configuration["Serilog:WriteTo:1:Args:serverUrl"];
+      }
+
+      Log.Logger = new LoggerConfiguration().ReadFrom
+        .Configuration(configuration)
+        .Enrich.WithProperty("Service", "UserService")
```
Comment (MariyaSin) [inline src/AchievementService/Program.cs]: UserService?????

File: src/AchievementService/Program.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+      }
+      finally
+      {
+        Log.CloseAndFlush();
+      }
+    }
+
+    public static IHostBuilder CreateHostBuilder(string[] args) =>
+      Host.CreateDefaultBuilder(args)
+          .UseSerilog()
```
Comment (MariyaSin) [inline src/AchievementService/Program.cs]: tabs

File: src/AchievementService/Startup.cs [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+    public void ConfigureServices(IServiceCollection services)
+    {
+      services.AddCors(options =>
+      {
+        options.AddPolicy(
+          CorsPolicyName,
+          builder =>
+          {
+            builder
+              //.WithOrigins(
```
Comment (MariyaSin) [inline src/AchievementService/Startup.cs]: you can remove from 156 to 162 lines

File: src/AchievementService/Startup.cs [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+
+      services.AddSingleton<IConnectionMultiplexer>(
+        x => ConnectionMultiplexer.Connect(redisConnStr));
+
+      services
+        .AddControllers(options =>
+        {
+          options.InputFormatters.Insert(0, GetJsonPatchInputFormatter());
+        }) // TODO check enum serialization from request without .AddJsonOptions()
+           //this will be used when all validation takes place on the pipeline
```
Comment (MariyaSin) [inline src/AchievementService/Startup.cs]: remove comments

File: src/AchievementService/AchievementService.csproj (hunk 10 above)
Comment (VladSpitsin) [reply src/AchievementService/AchievementService.csproj]: fixed

File: src/AchievementService/Program.cs (hunk 11 above)
Comment (VladSpitsin) [reply src/AchievementService/Program.cs]: fixed

File: src/AchievementService/Startup.cs (hunk 13 above)
Comment (VladSpitsin) [reply src/AchievementService/Startup.cs]: done

File: src/AchievementService/Startup.cs (hunk 14 above)
Comment (VladSpitsin) [reply src/AchievementService/Startup.cs]: done

File: src/AchievementService/Program.cs (hunk 12 above)
Comment (VladSpitsin) [reply src/AchievementService/Program.cs]: fixed

File: src/AchievementService/Program.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+      string seqServerUrl = Environment.GetEnvironmentVariable("seqServerUrl");
+      if (string.IsNullOrEmpty(seqServerUrl))
+      {
+        seqServerUrl = configuration["Serilog:WriteTo:1:Args:serverUrl"];
+      }
+
+      string seqApiKey = Environment.GetEnvironmentVariable("seqApiKey");
+      if (string.IsNullOrEmpty(seqApiKey))
+      {
+        seqApiKey = configuration["Serilog:WriteTo:1:Args:serverUrl"];
```
Comment (grinakate) [inline src/AchievementService/Program.cs]: maybe "Serilog:WriteTo:1:Args:apiKey"

Comment (VladSpitsin) [reply src/AchievementService/Program.cs]: fixed

File: src/AchievementService/Startup.cs [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+        });
+      }
+      else
+      {
+        services.Configure<RedisConfig>(Configuration.GetSection(RedisConfig.SectionName));
+      }
+
+      services.Configure<TokenConfiguration>(Configuration.GetSection("CheckTokenMiddleware"));
+      services.Configure<BaseServiceInfoConfig>(Configuration.GetSection(BaseServiceInfoConfig.SectionName));
+      services.Configure<BaseRabbitMqConfig>(Configuration.GetSection(BaseRabbitMqConfig.SectionName));
```
Comment (Cheephot) [inline src/AchievementService/Startup.cs:211]: add line

**Review by Cheephot — APPROVED** [review]

File: src/AchievementService/Startup.cs (hunk 16 above)
Comment (VladSpitsin) [reply src/AchievementService/Startup.cs:211]: fixed

File: src/AchievementService.Mappers/AchievementService.Mappers.csproj [hunk 17]
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <Import Project="$(ProjectDir)..\..\Common.Configurations.targets" />
+
+  <ItemGroup>
+    <PackageReference Include="SixLabors.ImageSharp" Version="1.0.3" />
+  </ItemGroup>
```
Comment (ndoriomedov) [inline src/AchievementService.Mappers/AchievementService.Mappers.csproj]: where is it used?

File: src/AchievementService/AchievementService.csproj [hunk 18]
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <Import Project="$(ProjectDir)..\..\Common.Configurations.targets" />
+
+  <ItemGroup>
+    <PackageReference Include="AspNetCore.HealthChecks.SqlServer" Version="5.0.3" />
+    <PackageReference Include="AspNetCore.HealthChecks.UI.Client" Version="5.0.1" />
```
Comment (ndoriomedov) [inline src/AchievementService/AchievementService.csproj]: update to latest rc versions (in manage nugets for project set "Include prerelease" check)

File: src/AchievementService/Properties/launchSettings.json [hunk 19]
```diff
@@ -0,0 +1,17 @@
+﻿{
+  "$schema": "http://json.schemastore.org/launchsettings.json",
+
+  "profiles": {
+    "AchievementService": {
+      "commandName": "Project",
+      "dotnetRunMessages": "true",
+      "launchBrowser": false,
+      "applicationUrl": "http://localhost:5000",
```
Comment (ndoriomedov) [inline src/AchievementService/Properties/launchSettings.json]: do we need this? @MariyaSin

File: src/AchievementService.Mappers/AchievementService.Mappers.csproj (hunk 17 above)
Comment (VladSpitsin) [reply src/AchievementService.Mappers/AchievementService.Mappers.csproj]: fixed

File: src/AchievementService/AchievementService.csproj (hunk 18 above)
Comment (VladSpitsin) [reply src/AchievementService/AchievementService.csproj]: done

File: src/AchievementService/Properties/launchSettings.json (hunk 19 above)
Comment (VladSpitsin) [reply src/AchievementService/Properties/launchSettings.json]: fixed

File: src/AchievementService.Data.Provider.MsSql.Ef/AchievementServiceDbContext.cs [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    public bool IsInMemory()
+    {
+      return Database.IsInMemory();
+    }
+
+    public async Task SaveAsync()
+    {
+      await SaveChangesAsync();
+    }
+
```
Comment (ekaterinaryzhenkova) [inline src/AchievementService.Data.Provider.MsSql.Ef/AchievementServiceDbContext.cs]: remove line

File: src/AchievementService.Models.Dto/Configurations/MemoryCacheConfig.cs [hunk 21]
```diff
@@ -0,0 +1,9 @@
+﻿namespace LT.DigitalOffice.AchievementService.Models.Dto.Configurations
+{
+  public record MemoryCacheConfig
+  {
+    public const string SectionName = "MemoryCache";
+
```
Comment (ekaterinaryzhenkova) [inline src/AchievementService.Models.Dto/Configurations/MemoryCacheConfig.cs:6]: remove line

Comment (VladSpitsin) [reply src/AchievementService.Models.Dto/Configurations/MemoryCacheConfig.cs:6]: no need here, I think

**Review by ekaterinaryzhenkova — APPROVED** [review]

**Review by MariyaSin — APPROVED** [review]

**Review by ndoriomedov — APPROVED** [review]

**Pull request merged**